lib/google/gax/config.rb

```diff
@@ -0,0 +1,127 @@
+# Copyright 2019, Google LLC
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are
+# met:
+#
+#     * Redistributions of source code must retain the above copyright
+# notice, this list of conditions and the following disclaimer.
+#     * Redistributions in binary form must reproduce the above
+# copyright notice, this list of conditions and the following disclaimer
+# in the documentation and/or other materials provided with the
+# distribution.
+#     * Neither the name of Google LLC nor the names of its
+# contributors may be used to endorse or promote products derived from
+# this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+module Google
+  module Gax
+    ##
+    # Config is a simple DSL for creating Configuration classes.
+    #
+    # @example
+    #   require "google/gax/config"
+    #
+    #   class SampleConfig
+    #     extend Google::Gax::Config
+    #
+    #     config_attr :name,   nil,         String, nil
+    #     config_attr :active, true,        true, false
+    #     config_attr :count,  nil,         Numeric, nil
+    #     config_attr :env,    :production, String, Symbol
+    #
+    #     def initialize parent_config = nil
+    #       @parent_config = parent_config unless parent_config.nil?
+    #       yield self if block_given?
+    #     end
+    #   end
+    #
+    #   config = SampleConfig.new
+    #
+    #   config.name             #=> nil
+    #   config.name = "thor"    #=> "thor"
+    #   config.name             #=> "thor"
+    #   config.name = :thor     # ArgumentError
+    #
+    module Config
+      ##
+      # Add configuration attribute methods to the configuratin class.
+      #
+      # @param [String, Symbol] key The name of the option
+      # @param [Object, nil] initial Initial value (nil is allowed)
+      # @param [Hash] opts Validation options
+      #
+      def config_attr name, default, *valid_values, &validator
```

Comment:
After looking at how default and parent interact below, I wonder if there's a way can do without default altogether, and just rely on specifying/setting a parent. Just a thought. I think it may make things conceptually simpler, but it may be more complicated to implement and still have a reasonable usage experience.

Reply:
I was kinda wondering the same. But having a default is super useful if we ever want to reset to it. e.g. a user changes the service host to something else, but wants to reset to the original value. If the config object doesn't maintain the default then I think we still want to set a default in a ruby constant. e.g. `Client::SERVICE_ADDRESS = "https://service.googleapis.com"`

Reply:
Yes, I agree we do want a "static" default, but the question is, where does it live? For example, the logic I'd expect to see is:

* Value set in the current config? Then return it.
* Value set in the parent config? Then return it. (And continue up the ancestry.)
* Got to the toplevel config and still nothing? Then return the default defined statically somewhere.

Currently that final default is declared in the `config_attr` declaration. But that brings up the question of *whose* `config_attr` declaration? It is currently possible for a config's parent to have a different class, which could have the same field names with different defaults. The current logic (as best as I can read it) says, start with the toplevel ancestor, and work backwards towards the most detailed override, and take the first non-nil default. (i.e. defaults are chosen in reverse order from values.) I don't think this is intuitive behavior. We could solve this by enforcing that a parent config must have the same class (and thus the same fields and defaults) as a child config. But I'm not sure if you intended to allow duck-typing in the parent.

Reply:
This PR is meant to be a drop-in replacement for #195, but one thing that using this DSL does that the previous Configuration implementation doesn't is allow us to use Ruby inheritance. This way, a default can be provided whenever an attribute is defined, and new configs can inherit from an existing config class. So, for example, I hope to end up with something like following if we go with this approach:

```ruby
class Google::Cloud::Vision::Configure
  extend Google::Gax::Config

  config_attr :timeout, 60, Numeric
end

class Google::Cloud::Vision::V1::ImageAnnotator::Client::Configure < Google::Cloud::Vision::Configure
  config_attr :host, "vision.googleapis.com", String
  config_attr :port, 443, Integer
  # timeout is already configured
end
```

But we still have the instance parent/derived interplay:

```ruby
parent_config = Google::Cloud::Vision::V1::ImageAnnotator::Client::Configure.new do |config|
  config.timeout = 90
end

derived_config = Google::Cloud::Vision::V1::ImageAnnotator::Client::Configure.new(parent_config) do |config|
  config.timeout = 90
end

parent_config.timeout #=> 90
derived_config.timeout #=> 120

derived_config.timeout = nil # reset to parent config
derived_config.timeout #=> 90
parent_config.timeout #=> 90

parent_config.timeout = nil # reset to default
parent_config.timeout #=> 60
derived_config.timeout #=> 60
```